core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -614,18 +614,24 @@ private[spark] class ExecutorAllocationManager(
             stageAttemptToNumRunningTask -= stageAttempt
           }
         }
-        // If the task failed, we expect it to be resubmitted later. To ensure we have
-        // enough resources to run the resubmitted task, we need to mark the scheduler
-        // as backlogged again if it's not already marked as such (SPARK-8366)
-        if (taskEnd.reason != Success) {
-          if (totalPendingTasks() == 0) {
-            allocationManager.onSchedulerBacklogged()
-          }
-          if (taskEnd.taskInfo.speculative) {
-            stageAttemptToSpeculativeTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
-          } else {
-            stageAttemptToTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
-          }
+
+        if (taskEnd.taskInfo.speculative) {
+          stageAttemptToSpeculativeTaskIndices.get(stageAttempt).foreach {_.remove{taskIndex}}
+          stageAttemptToNumSpeculativeTasks(stageAttempt) -= 1
+        }
+
+        // If the task failed (not intentionally killed), we expect it to be resubmitted later. To
+        // ensure we have enough resources to run the resubmitted task, we need to mark the
+        // scheduler as backlogged again if it's not already marked as such (SPARK-8366)
+        taskEnd.reason match {
+          case Success | _: TaskKilled =>
```

Comment:
Why do we want to include `TaskKilled` here? It's not quite straight forward to me

Reply:
Inside the the brackets, there are two things:
```
if (totalPendingTasks() == 0) {
    allocationManager.onSchedulerBacklogged()
}
```
This one is straightforward. If a task is intentionally killed, then we don't expect this task to be resubmitted again, and we don't need to mark the scheduler as backlogged.
```
if (!taskEnd.taskInfo.speculative) {
    stageAttemptToTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
}
```
If a non-speculative task is intentionally killed, it means the speculative task has succeeded, and no further task of this task index will be resubmitted. In this case, the task index is completed and we shouldn't remove it from `stageAttemptToTaskIndices`. Otherwise, we will have a pending non-speculative task for the task index.

Reply:
I think it would be good to add a comment under the taskkilled explaining this